### PR TITLE
Typed Builder implementation Part 1

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality

### CODE REVIEW

**Comment by Kukovec** [comment]
As per our discussion at Friday's meeting, I've added scope-validation to the builder. 
There's 2 levels to building: 
  1. "raw" building, which takes `n` (presumably well typed) arguments and constructs an IR node, with a given operator, and performs toplevel type computation for the resulting expression. For example, given arguments `a1: Set(Int), a2: Set(Int)`, the raw `_union` should produce a `OperEx(TlaSetOper.union, a1, a2)(Typed(Set(IntT1())))`, because `\union` is an operator with a known signature `(Set(t), Set(t)) => Set(t)`, which can be unified with the arguments `(Set(Int), Set(Int))`. Note that raw building does not guarantee total type correctness, e.g. if 
  ```a1 = OperEx(TlaSetOper.enum, NameEx("a")(Untyped()))(Typed(SetT1(IntT1())))``` 
nor scope correctness, e.g. in `forall(x: Int, S: Set(Int), x: Bool)`, as the signature of `forall` is `(t, Set(t), Bool) => Bool` and agnostic to actual names.

2. Scoped building, which operates over `BuilderWrapper` expressions, which are `State` values carrying the scope as internal state, producing `TlaEx` on demand. Scoped building guarantees that, as long as all IR components are built with the `ScopedBuilder`, the final expression will be both type-correct and scope-consistent.


At his point in time, you can ignore: 
  - Applicative
  - BoolOp
  - FunOp
  - TypeCalculatingBuilder

Comments on the draft would be appreciated at this time @konnov @thpani @shonfeder @gabrielamafra

**Comment by Kukovec** [comment]
This is still missing tests for `TlaSetOper` methods, but that shouldn't be a blocker for reviewing.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:88]: You don't actually know what this is an alternative to when you're reading the Scaladoc.
I suggest copying the above doc comment and linking the other method with `@see`.

**Review by thpani — COMMENTED** [review]
Nicely organized, clean code, I really like this! ✨
I have a few comments inline.

LGTM from my side; I'm only not approving this so others have time to take a look.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala:25]: Future-proofing. For a concrete example, I'm considering adding automatic coverage analysis (i.e. which "branches" assign to which variables) to the builder (very far off in the future), so that when you build an expression it is automatically correct w.r.t. assignments. In that case, I'd extend `MetaInfo` with e.g. another map, but `markAsBound` wouldn't have to care about other components of `MetaInfo`.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/SignatureGenerator.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/SignatureGenerator.scala:10]: I think I noticed this one myself after I had already pushed, this is leftover from when the generator did more things that are now delegated to other classes.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/SignatureGenerator.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/SignatureGenerator.scala:69]: This method used to be part of the package object, so it had to be stateless. I guess we can have a persistent `Unifier` now, but the performance diff is probably marginal either way.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala:29]: I don't remember, I think I just went with the conventions from the 1st draft. But it is strange, now that you point it out. I'll uppercase them.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:20]: To me, `require` is a necessary evil in methods which both a) take a variable number of arguments, and b) impose constraints only on _some_ of the arguments (e.g. every other arg. has to be a `NameEx` or a string `ValEx`).
In your case, you can do 
```scala
val h +: tail = args
enumSet(h, tail:_*)
```
instead of `enumSet(args:_*)`, so it shouldn't be too bad.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala
Comment (thpani) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala:25]: 👍

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (thpani) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:20]: I won't block on this, but I would like to hear others' thoughts.

Comment (thpani) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:20]: Pinging @shonfeder @konnov @gabrielamafra

**Review by thpani — APPROVED** [review]
This has been sitting around for some time; if it doesn't get more looks I might as well LGTM it.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala:50]:
```suggestion
  type TBuilderPiece = InternalState[BuilderReturn]
```

Do the users of your API care about the fact that this is a wrapper? It's important that this is a piece of work to be done by the builder. So I would called it a "piece", a "part", etc.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala:49]: Since the users will see this type in their IDE quite often, it needs a good description, not a comment about its internal mechanics.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala:54]:
```suggestion
  type TBuilderNamePiece = InternalState[NameEx]
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala:57]:
```suggestion
  type TBuilderValPiece = InternalState[ValEx]
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawLeafBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawLeafBuilder.scala:12]:
```suggestion
trait RawLiteralAndNameBuilder {
```

I would prefer a descriptive name, especially if it's used only internally.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/ProtoBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/ProtoBuilder.scala:24]:
```suggestion
  protected def buildBySignatureLookup(oper: TlaOper, args: TlaEx*): TlaEx =
```

I think `simpleInstruction` is only meaningful in the eyes of the author. This method may construct operators that are not simple to me. Moreover, I (as a method user) don't really care whether it was simple to implement or not. It is probably important that such an expression is essentially constructed by a table lookup.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawArithmeticBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawArithmeticBuilder.scala:12]:
```suggestion
trait UnsafeArithmeticBuilder extends ProtoBuilder {
```

Again, "raw" does not tell much to the reader. I would call it "unsafe", because it may construct ill-typed expressions. Hence, the API user should not even look at this class.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawBoolBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawBoolBuilder.scala:12]:
```suggestion
trait UnsafeBoolBuilder extends ProtoBuilder {
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala:17]:
```suggestion
trait UnsafeSetBuilder extends ProtoBuilder {
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala:49]: I found it annoying that the builder differentiates between the first parameter and the other parameters. So I would prefer just `TlaEx*` or maybe `(NameEx, TlaEx)*`, which you can nicely write as `n1 -> e1, n2 -> e2, ...`. Perhaps, shapeless can help us in doing static checks to make sure that the list is non-empty. @shonfeder ?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala:60]: Same here. I would prefer not to distinguish the first field from the other fields.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/raw/RawSetBuilder.scala:113]: Same here. There should be no difference between the first pair and the other pairs.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/LeafBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/LeafBuilder.scala:15]:
```suggestion
trait LiteralAndNameBuilder extends RawLeafBuilder {
```

Names should reflect the use, not the implementation or another unnecessary detail.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:20]: Yeah, I also wrote above that I did not like that about the old builder. If we want to enforce some properties of a sequence at compile-time, we should not put this burden on the user. And @shonfeder mentioned shapeless.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:60]: Same here. I would prefer a uniform interface. Perhaps, a sequence of pairs? Scala's `->` can be very nicely applied here.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:81]: Wanted: uniform treatment of all arguments, e.g., `(ValWrapper, BuilderWrapper)*`.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:92]: Wanted: uniform treatment of all arguments, e.g., (String, BuilderWrapper)*.

**Review by konnov — COMMENTED** [review]
Another batch of comments. I like your approach! My main concern is that the implementation details sneak into the naming and the methods too much. Hence, most of the changes are purely syntactic.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestArithmeticBuilder.scala
Comment (konnov) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestArithmeticBuilder.scala:13]:
```suggestion
    val args = Seq(builder.int(10), builder.int(20))
```

I don't like riddles in tests :)

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestBoolBuilder.scala
Comment (konnov) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestBoolBuilder.scala:25]:
```suggestion
    val illTypedY: BuilderWrapper = builder.str("a")
```

**Review by konnov — COMMENTED** [review]
Just a few comments about the tests. End of my comments :)

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala:1]: "computation"

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/SignatureHandler.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/SignatureHandler.scala:16]: `-Factory` maybe, but not `Signature-`, rather `TypeComputation-`, since it imports signatures and makes computations.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala:50]: `-Instruction`? Since an `InternalState[TlaEx]` defines how to build `TlaEx` on demand?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:20]: Ok, so the general sentiment seems to be to enforce minimal arity via runtime checks, so I'll go ahead and change those methods. 
Just one question, regarding maps (or methods with asymmetric args in general): do you prefer
`builder.map(e: T, xsAndSs: T*)` + `require(P(xsAndSs))` or `builder.map(eAndXsAndSs: T*)` + `require(P(eAndXsAndSs))`. Here, `e` and `xsAndSs` play different roles, and it's very unlikely that `eAndXsAndSs` would be programmatically constructed, since the components have different roles and shapes (unlike `xsAndSs`, which is uniform in shape).
I'm more in favor of the former, but I'd like to hear your opinions.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:60]: I can make a `BuilderWrapper*` variant and a `(A,B)*` variant, like I do with record sets.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestArithmeticBuilder.scala
Comment (Kukovec) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestArithmeticBuilder.scala:13]: I'll hold off on these changes, because I'm working on the generalzied testing framework where this should be parameterized by arity anyway.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestArithmeticBuilder.scala
Comment (Kukovec) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecmp/TestArithmeticBuilder.scala:20]: See above

**Review by Kukovec — COMMENTED** [review]
I'll incorporate @konnov 's naming suggestions, but not with GH's "commit suggestions", since they need refactoring to work.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:20]: I would definitely prefer `builder.map(e: T, xsAndSs: T*)`

Comment (thpani) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala:20]: +1

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/package.scala:44]: Hmm, this is a transitional issue at best, since there's not going to be another builder after the overhaul. But I'll change it for now.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala:1]: This would be clearer as

```suggestion
package at.forsyte.apalache.tla.typecomp
```

**Review by konnov — APPROVED** [review]
Nothing catches my eye! Looks great!

**Comment by thpani** [comment]
I can't do a delta review here, because the package renaming breaks the diff. Also this PR has become so large that I don't want to do another full review.
IMO we should've merged this at some point and followed up with another PR to add the tests.

**Pull request merged**
